unity/Assets/Scripts/ActionDispatcher.cs

```diff
@@ -167,75 +167,74 @@ public static class ActionDispatcher {
         Dictionary<string, List<MethodInfo>> methodDispatchTable = getMethodDispatchTable(targetType);
-        if (!methodDispatchTable.ContainsKey(action))
-        {
+        if (!methodDispatchTable.ContainsKey(action)) {
             List<MethodInfo> methods = new List<MethodInfo>();
 
             List<Type> hierarchy = new List<Type>();
+
             // not completely generic
             Type ht = targetType;
-            while (ht != typeof(object))
-            {
+            while (ht != typeof(object)) {
                 hierarchy.Add(ht);
                 ht = ht.BaseType;
             }
 
-            foreach (MethodInfo mi in getMethods(targetType))
-            {
-                if (mi.ReturnType != typeof(void) || mi.Name != action)
-                {
+            foreach (MethodInfo mi in getMethods(targetType)) {
+                if (mi.ReturnType != typeof(void) || mi.Name != action) {
                     // We only allow dispatching to public void methods
                     continue;
                 }
                 bool replaced = false;
+
                 // we do this to handle the case of a child method hiding a method in the parent
                 // in which case both methods will show up.  This happens if virtual, override or new
                 // are not used
                 ParameterInfo[] sourceParams = mi.GetParameters();
 
-                for (int j = 0; j < methods.Count && !replaced; j++)
-                {
+                for (int j = 0; j < methods.Count && !replaced; j++) {
                     bool signatureMatch = true;
                     ParameterInfo[] targetParams = methods[j].GetParameters();
                     int minCommon = Math.Min(sourceParams.Length, targetParams.Length);
-                    for (int k = 0; k < minCommon; k++)
-                    {
-                        if (sourceParams[k].ParameterType != targetParams[k].ParameterType)
-                        {
+                    for (int k = 0; k < minCommon; k++) {
+                        if (sourceParams[k].ParameterType != targetParams[k].ParameterType) {
                             signatureMatch = false;
                         }
                     }
 
-                    if (sourceParams.Length > targetParams.Length && !sourceParams[minCommon].HasDefaultValue)
-                    {
-                        signatureMatch = false;
-                    }
-                    else if (targetParams.Length > sourceParams.Length && !targetParams[minCommon].HasDefaultValue)
-                    {
-                        signatureMatch = false;
+                    if (sourceParams.Length != targetParams.Length) {
+                        // Some methods, like teleport full have
+                        // add both required args and optional args to the end.
+                        for (int i = minCommon - 1; i < sourceParams.Length; i++) {
```

Comment:
Should `i` start at `minCommon` or `minCommon-1`? It seems that the old code only checked starting at `minCommon`, was this an error?

Reply:
This is a bug fix and should use `minCommon - 1`. Consider if it checks:

`sourceParams`:

0. `position`
1. `standing`
2. `forceAction` optional

and

`targetParams`:

0. `position`
1. `forceAction` optional

Here, `minCommon=2`, so we should start checking the `sourceParams` at `standing`, since we don't have to order it like:
`sourceParams`:

0. `position`
1. `forceAction` optional
2. `standing`

Reply:
I'm not sure I understand the need for this change.  If we consider your first example:

0. `position`
1. `standing`
2. `forceAction` optional

vs

0. `position`
1. `forceAction` optional

The first loop that iterates up through minCommon will set signatureMatch=False, because it will iterate up to element 1 (standing vs forceAction) and these will fail the type check.  

The purpose of this bit of code is to effectively allow overriding actions that have common initial set of parameters.  So something like:

MyAction(float, float) could be overridden by MyAction(float,float, bool=True) in a subclass.  The idea is that this behavior would offer the least amount of surprise to someone editing actions across the hierarchy.

Reply:
The above loop was only checking for the type, so, since `standing` is a boolean and `forceAction` is a boolean, so `signatureMatch` never actually got set to `false`.

Reply:
Ahh, I see. You were getting a false positive for signature match when you shouldn't have.  Okay, then the use case that I think will break with your change is this:

source:
0. `param1` (float)
1. `param2` (float)
2. `param3` (bool=True) optional

target:
0. `param1` (float)
1. `param2` (float)

The first check will succeed since the minCommon number of params have identical types.  The new block will get entered since the number of params is different.  Within here the signatureMatch would get set to false since both source[minCommon-1] and target[minCommon-1] don't have default values, but we do want this to be considered a match since if both methods are available to be dispatched to its ambiguous about which one to pick if you pass just param1 and param2, which is why we allow this replacement if the source method was defined lower in the class hierarchy.

Its my fault for never writing a test for this (I am going to add one on master).

@Lucaweihs, what are you thoughts about 1. the existing override logic and 2. making the override logic sensitive to the names and types instead of just the types, which should make it trivial to support the case above.

Reply:
Okay, then the use case that I think will break with your change is this:

source:

0. `param1` (float)
1. `param2` (float)
2. `param3` (bool=True) optional

target:

0. `param1` (float)
1. `param2` (float)

The first check will succeed since the minCommon number of params have identical types.  The new block will get entered since the number of params is different.  Within here the signatureMatch would get set to false since both source[minCommon-1] and target[minCommon-1] don't have default values, but we do want this to be considered a match since if both methods are available to be dispatched to its ambiguous about which one to pick if you pass just param1 and param2.  In the signature match block we check whether the source and target params match in length and then throws an AmbiguousActionException.  The main purpose of all this signatureMatching is to 1. look for more specific matches in the class hierarchy (with the same number of params). 2. Look for ambiguous matches such as this one.

Reply:
I originally wrote the signatureMatch logic to try to resemble the behavior that csharp has with respect to method dispatching, but we dispatch exclusively using parameterNames.  I wonder if we just switched to the following logic whether it would be sufficient.  

a = {source.paramNames}
b = {target.paramNames}

c = a - b 

If the parameterNames in set c all have defaults, then we consider source and target a match.  I think this satisfies both of the requirements above.

Reply:
For now, I've just gone back to what it was, because I moved Teleport and TeleportFull to protected methods from the base.

I agree that the suggested set differences approach is the way to go.